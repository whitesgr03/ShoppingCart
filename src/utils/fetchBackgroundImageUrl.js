import { initialStorage } from "../firebase-config";

const storage = initialStorage();

const fetchBackgroundImageUrl = async page => {
	const ref = storage.ref(`images/${page}/background.jpg`);
	return await ref.getDownloadURL();
};

export { fetchBackgroundImageUrl as default };
