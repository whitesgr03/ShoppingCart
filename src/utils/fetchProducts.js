import { initialFirestore } from "../firebase-config";

const firestore = initialFirestore();

const fetchProducts = async () => {
	const productRef = firestore.collection("products");
	const result = await productRef.get();

	return result.empty
		? (() => {
				throw result;
		  })()
		: result.docs.map(item => {
				const { name, price, url } = item.data();
				return {
					id: item.id,
					name,
					url,
					price,
				};
		  });
};

export { fetchProducts as default };
