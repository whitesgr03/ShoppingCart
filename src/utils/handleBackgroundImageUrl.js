import { initialStorage } from "../firebase-config";

const getBackgroundImageUrl = async page => {
	const storage = initialStorage();
	const ref = storage.ref(`images/${page}/background.jpg`);

	try {
		const url = await ref.getDownloadURL();

		return {
			message: "Get background-image url success.",
			success: true,
			data: url,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

export { getBackgroundImageUrl as default };
