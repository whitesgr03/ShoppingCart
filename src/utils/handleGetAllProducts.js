import { initialFirestore } from "../firebase-config";

const handleGetAllProducts = async () => {
	const firestore = initialFirestore();
	const productsRef = firestore.collection("products");

	const result = await productsRef.get();

	const allProductsResult = result.empty
		? []
		: result.docs.map(item => ({
				id: item.id,
				...item.data(),
		  }));

	return allProductsResult;
};

export default handleGetAllProducts;
