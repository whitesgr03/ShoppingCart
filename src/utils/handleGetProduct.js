import { initialFirestore } from "../firebase-config";

const handleGetProduct = async productId => {
	const firestore = initialFirestore();
	const productRef = firestore.collection("products").doc(productId);

	const result = await productRef.get();

	const productResult = {
		id: result.id,
		...result.data(),
	};

	return productResult;
};

export default handleGetProduct;
