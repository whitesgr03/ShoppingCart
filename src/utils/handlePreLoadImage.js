const preLoadImage = url =>
	new Promise(resolve => {
		const img = new Image();
		img.addEventListener("load", () => resolve(url), { once: true });
		img.src = url;
	});
	
export default preLoadImage;
