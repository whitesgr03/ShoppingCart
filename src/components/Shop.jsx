import { useState, useEffect } from "react";
import { Outlet, Link, useLocation, useOutletContext } from "react-router-dom";

import Icon from "@mdi/react";
import { mdiMagnify, mdiLoading } from "@mdi/js";

import { fetchResource } from "../utils/utils";

const Shop = () => {
	const { onAddItem, onToggleModal } = useOutletContext();
	const { pathname, key: currentKey } = useLocation();

	const isShopRoute = pathname === "/shop";

	const [products, setProducts] = useState(null);
	const [filterText, setFilterText] = useState("");
	const [previousKey, setPreviousKey] = useState(currentKey);

	useEffect(() => {
		const handleFetch = async () => {
			setProducts(await fetchResource());
		};
		handleFetch();
	}, []);

	useEffect(() => {
		const isKeyChanged = previousKey !== currentKey;

		const handleClearFilterText = () => {
			setFilterText("");
			setPreviousKey(currentKey);
		};

		isShopRoute && isKeyChanged && handleClearFilterText();
	});

	return (
		<div data-testid="shop" className="shop">
			{products ? (
				<>
					<Navbar
						isShopRoute={isShopRoute}
						filterText={filterText}
						onFilterTextChange={setFilterText}
					/>

					<Outlet
						context={{
							products,
							filterText,
							onAddItem,
							onToggleModal,
						}}
					/>
				</>
			) : (
				<div data-testid="loading" className="loading">
					<Icon path={mdiLoading} spin={1} size={3} />
					{"Loading..."}
				</div>
			)}
		</div>
	);
};

const Navbar = ({
	isShopRoute, 
	filterText,
	onFilterTextChange
}) => {
	const [active, setActive] = useState(false);

	const handleSubmit = e => {
		e.preventDefault();
		onFilterTextChange(...new FormData(e.target).values());
	};

	const handleChange = e => {
		onFilterTextChange(e.target.value);
	};

	const handleActiveSearchBarBorder = () => {
		setActive(true);
	};

	const handleDeActivateSearchBarBorder = () => {
		setActive(false);
	};

	return (
		<div className="navigation" data-testid="navigation">
			<ul className="category">
				<li>
					<Link to="/shop">Shop</Link>
				</li>
				<li> / </li>
				<li>
					<Link to="/shop">Category</Link>
				</li>
			</ul>
			{isShopRoute && (
				<form
					data-testid="searchBar"
					className={`searchBar ${active ? "active-border" : ""}`}
					onFocus={handleActiveSearchBarBorder}
					onBlur={handleDeActivateSearchBarBorder}
					onSubmit={handleSubmit}
				>
					<div className="search-border">
						<input
							data-testid="search"
							type="search"
							name="search"
							placeholder="search..."
							value={filterText}
							onChange={handleChange}
						/>
					</div>
					<button type="submit">
						<Icon path={mdiMagnify} />
					</button>
				</form>
			)}
		</div>
	);
};

const Products = () => {
	const { products, filterText } = useOutletContext();

	const filterProducts =
		filterText === ""
			? products
			: products.filter(
					item =>
						item.name
							.toLowerCase()
							.indexOf(filterText.toLowerCase().trim()) !== -1
			  );

	const productList =
		products.length > 0 ? (
			filterProducts.map(product => (
				<div className="item" key={product.id}>
					<Link
						className="image-link"
						to={`/shop/${product.id}`}
						state={{ product }}
					>
						<img src={product.url} alt={product.name} />
					</Link>
					<div className="info">
						<div className="title">
							<Link
								className="text-link"
								to={`/shop/${product.id}`}
								state={{ product }}
							>
								{product.name}
							</Link>
						</div>
						<div className="price">${product.price.toFixed(2)}</div>
					</div>
				</div>
			))
		) : (
			<div>{"loading..."}</div>
		);

	return (
		<div
			data-testid="products"
			className={`products ${products.length > 0 ? "" : "loading"}`}
		>
			{productList}
		</div>
	);
};

const ProductInfo = () => {
	const { state } = useLocation();

	const { onAddItem, onToggleModal } = useOutletContext();

	const maxOptions = 10;

	const options = Array.from({ length: maxOptions })
		.fill([])
		.map((_, i) => (
			<option key={i} value={i + 1}>
				{i + 1}
			</option>
		));

	const handleSubmit = e => {
		e.preventDefault();

		const formData = new FormData(e.target).entries();

		const entriesNumeric = [...formData].map(([key, value]) =>
			key === "quantity" ? [key, Number(value)] : [key, value]
		);

		const newProduct = {
			...state.product,
			...Object.fromEntries(entriesNumeric),
		};

		onAddItem(newProduct);
		onToggleModal("showAlert");
	};

	return state ? (
		<div className="productInfo">
			<div className="image">
				<img src={state.product.url} alt={state.product.name} />
			</div>
			<div className="info">
				<div className="description">
					<h2 className="title">{state.product.name}</h2>
					<p>
						Lorem ipsum dolor sit amet consectetur adipisicing elit.
						Magnam itaque nisi ipsa pariatur perspiciatis nemo.
					</p>
					<p className="price">${state.product.price.toFixed(2)}</p>
				</div>

				<form onSubmit={handleSubmit}>
					<label htmlFor="quantity">
						Quantity
						<select name="quantity" id="quantity" defaultValue={1}>
							{options}
						</select>
					</label>
					<button className="slide" type="submit">
						Add to Cart
					</button>
				</form>
			</div>
		</div>
	) : (
		<div data-testid="productError" className="productError">
			<div className="error">
				<p>Our apologies, there has been an error.</p>
				<p>Please come back later or return to the previous page.</p>
			</div>
		</div>
	);
};

export { Shop as default, Navbar, Products, ProductInfo };
