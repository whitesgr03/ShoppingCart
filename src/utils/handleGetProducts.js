import { initialFirestore } from "../firebase-config";
import handlePreLoadImage from "./handlePreLoadImage";

const handleGetProducts = async () => {
	const firestore = initialFirestore();
	const productRef = firestore.collection("products");

	const result = await productRef.get();

	await Promise.all(
		result.docs.map(item => {
			const obj = { ...item.data() };
			return handlePreLoadImage(obj.url);
		})
	);

	const productsResult = result.docs.map(item => ({
		id: item.id,
		...item.data(),
	}));

	return productsResult;
};

export default handleGetProducts;
