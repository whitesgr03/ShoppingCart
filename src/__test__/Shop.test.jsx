import { render, screen, waitFor } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router-dom";

import userEvent from "@testing-library/user-event";

import Shop, { Navbar, Products, ProductInfo } from "../components/Shop";

import { fetchResource } from "../utils/utils";


jest.mock("react-router-dom", () => ({
	...jest.requireActual("react-router-dom"),
	useLocation: jest
		.fn()
		.mockReturnValueOnce({})
		.mockReturnValue({
			state: {
				product: { id: 0, name: "fake", url: "../", price: 19.9 },
			},
		}),
	useOutletContext: jest
		.fn()
		.mockReturnValueOnce({
			products: [],
			filterText: "",
		})
		.mockReturnValueOnce({
			products: [
				{
					id: 0,
					name: "fakeBag",
					url: "../",
					price: 19.9,
				},
				{
					id: 1,
					name: "fakePants",
					url: "../",
					price: 19.9,
				},
			],
			filterText: "Bag",
		})
		.mockReturnValue({
			products: [
				{
					id: 0,
					name: "fakeBag",
					url: "../",
					price: 19.9,
				},
				{
					id: 1,
					name: "fakePants",
					url: "../",
					price: 19.9,
				},
			],
			filterText: "",
		}),
}));

jest.mock("../utils/handleResource", () => ({
	...jest.requireActual("../utils/handleResource"),
	fetchResource: jest.fn(),
}));

describe("Renders Navbar Component", () => {
	it("Should return Navbar DOM", () => {
		const routes = [
			{
				path: "/",
				element: <Navbar />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const actual = screen.getByTestId("navigation");

		expect(actual).toHaveClass("navigation");
	});
	it("Should return searchBar DOM", () => {
		const mockIsShopRoute = true;

		const routes = [
			{
				path: "/",
				element: <Navbar isShopRoute={mockIsShopRoute} />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const actual = screen.getByTestId("searchBar");

		expect(actual).toHaveClass("searchBar");
	});

	it("Should return filterText with click event", async () => {
		const user = userEvent.setup();

		const mockIsShopRoute = true;
		const mockFilterText = "Bag";
		const mockOnFilterTextChange = jest.fn();

		const routes = [
			{
				path: "/",
				element: (
					<Navbar
						isShopRoute={mockIsShopRoute}
						filterText={mockFilterText}
						onFilterTextChange={mockOnFilterTextChange}
					/>
				),
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const button = screen.getByRole("button");

		await user.pointer({ keys: "[MouseLeft]", target: button });

		expect(mockOnFilterTextChange.mock.calls[0][0]).toEqual(mockFilterText);
	});

	it("Should return filterText with change event", async () => {
		const user = userEvent.setup();

		const mockIsShopRoute = true;
		const mockOnFilterTextChange = jest.fn();

		const routes = [
			{
				path: "/",
				element: (
					<Navbar
						isShopRoute={mockIsShopRoute}
						onFilterTextChange={mockOnFilterTextChange}
					/>
				),
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const input = screen.getByTestId("search");

		const mockFilterText = "Bag";

		await user.type(input, mockFilterText);

		expect(mockOnFilterTextChange).toBeCalledTimes(3);

		expect(mockOnFilterTextChange.mock.lastCall[0]).toEqual(mockFilterText);
	});
});

describe("Renders Products Component", () => {
	it("Should show loading with empty products", () => {
		const routes = [
			{
				path: "/",
				element: <Products />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const actual = screen.getByTestId("products");

		expect(actual).toHaveClass("loading");
	});

	it("Should show products matching the search text", () => {
		const routes = [
			{
				path: "/",
				element: <Products />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const products = screen.getByTestId("products");

		expect(products).not.toHaveClass("loading");

		expect(screen.queryByText("fakePants")).not.toBeInTheDocument();

		const actual = screen.getByRole("img");

		expect(actual).toHaveAttribute("src", "../");

		expect(screen.getByText("fakeBag")).toBeInTheDocument();
		expect(screen.getByText("$19.90")).toBeInTheDocument();
	});

	it("Should show all products without searchText", () => {
		const routes = [
			{
				path: "/",
				element: <Products />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const products = screen.getByTestId("products");

		expect(products).not.toHaveClass("loading");

		expect(screen.getByText("fakePants")).toBeInTheDocument();
		expect(screen.getByText("fakeBag")).toBeInTheDocument();
	});
});

describe("Renders ProductInfo Component", () => {
	it("Should return ProductInfo DOM without state", () => {
		const routes = [
			{
				path: "/",
				element: <ProductInfo />,
			},
		];
		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});
		render(<RouterProvider router={router} />);

		const actual = screen.getByTestId("productError");

		expect(actual).toHaveClass("productError");
	});

	it("Should return ProductInfo DOM with state", () => {
		const routes = [
			{
				path: "/",
				element: <ProductInfo />,
			},
		];
		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});
		render(<RouterProvider router={router} />);

		expect(screen.getByRole("img")).toHaveAttribute("src", "../");

		expect(screen.getByText("fake")).toHaveClass("title");

		expect(screen.getByText("$19.90")).toHaveClass("price");
	});
});

describe("Renders Shop Component", () => {
	it("Should return Shop DOM with fetch", async () => {
		const routes = [
			{
				path: "/",
				element: <Shop />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const actual = screen.getByTestId("shop");

		expect(actual).toHaveClass("shop");

		await waitFor(() => {
			expect(fetchResource).toBeCalledTimes(1);
		});
	});
});
