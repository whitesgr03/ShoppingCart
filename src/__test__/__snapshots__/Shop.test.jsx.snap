// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Shop Component Should return Navbar Component 1`] = `
<div>
  <div
    class="navigation"
  >
    <ul
      class="category"
    >
      <li>
        Home /
      </li>
      <li>
        Category /
      </li>
      <li>
        Top
      </li>
    </ul>
    <form
      class="search "
    >
      <div
        class="search-border"
      >
        <input
          placeholder="search..."
          type="search"
        />
      </div>
      <button
        type="submit"
      >
        <svg
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z"
            style="fill: currentColor;"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;
